vault/src/execution.rs

```diff
@@ -465,82 +507,94 @@ pub async fn execute_open_requests(
+            // fail to process requests. On the other hand, if we ignore transactions,
+            // we risk double paying. We choose to restart only for network errors,
+            // since these are not expected to be persistent. Other errors could be
+            // persistent, so we keep going.
+            let tx = match result {
+                Ok(x) => x,
+                Err(e) if e.is_transport_error() => {
+                    return Err(e.into());
+                }
+                Err(e) => {
+                    tracing::warn!("Failed to process transaction: {}", e);
+                    continue;
+                }
+            };
 
-    // iterate through transactions in reverse order, starting from those in the mempool, and
-    // gracefully fail on encountering a pruned blockchain
-    let mut transaction_stream = bitcoin::reverse_stream_transactions(&read_only_btc_rpc, btc_start_height).await?;
-    while let Some(result) = transaction_stream.next().await {
-        if rate_limiter.check().is_ok() {
-            // give the outer `select` a chance to check the shutdown signal
-            tokio::task::yield_now().await;
+            // get the request this transaction corresponds to, if any
+            if let Some(request) = get_request_for_btc_tx(&tx, &open_requests) {
+                // remove request from the hashmap
+                open_requests.retain(|&key, _| key != request.hash);
+
+                // start a new task to (potentially) await confirmation and to execute on the parachain
+                // make copies of the variables we move into the task
+                create_payment_worker(
+                    shutdown_tx.clone(),
+                    parachain_rpc.clone(),
+                    vault_id_manager.clone(),
+                    request,
+                    tx.txid(),
+                    num_confirmations,
+                    auto_rbf,
+                );
+            }
         }
+    }
 
-        // When there is an error, we have to make a choice. Either we restart or
-        // continue processing. Both options have their risks: if we restart, it's
-        // possible that the vault will never manage to start up, causing us to
-        // fail to process requests. On the other hand, if we ignore transactions,
-        // we risk double paying. We choose to restart only for network errors,
-        // since these are not expected to be persistent. Other errors could be
-        // persistent, so we keep going.
-        let tx = match result {
-            Ok(x) => x,
-            Err(e) if e.is_transport_error() => {
-                return Err(e.into());
+    // All requests remaining in the hashmap did not have a bitcoin payment yet
+    // or were not found in the stream, check if we can fetch by id directly
+    // or just pay and execute all requests individually
+    for (id, request) in open_requests {
+        // try finding transaction directly (if supported)
```

Comment:
Let's move this to the `else` branch of `is_full_node`?

Reply:
We need the `request` so if I moved that to the `else` branch we would need to do a separate iteration over the `open_requests`. This structure is why I opted to return `Ok(None)` in the bitcoin-core implementation